include/deal.II/lac/cuda_sparse_matrix.h

```diff
@@ -0,0 +1,319 @@
+    /**
+     *  Declare type for container size.
+     */
+    typedef unsigned int size_type;
+
+    /**
+     * Type of the matrix entries.
+     */
+    typedef Number value_type;
+
+    /**
+     * Declare a type that holds real-valued numbers with the same precision
+     * as the template argument to this class.
+     */
+    typedef Number real_type;
+
+    /**
+     * @name Constructors and initialization
+     */
+    //@{
+    /**
+     * Constructor. Initialize the matrix to be empty, without any structure,
+     * i.e., the matrix is not usable at all. This constructor is therefore
+     * only useful for matrices which are members of a class.
+     *
+     * You have to initialize the matrix before usage with reinit.
+     */
+    SparseMatrix();
+
+    /**
+     * Constructor. Takes a cuSPARSE handle and a sparse matrix on the host.
+     * The sparse matrix on the host is copied on the device and the elements
+     * are reordered according to the format supported by cuSPARSE.
+     */
+    SparseMatrix(cusparseHandle_t handle,
+                 const ::dealii::SparseMatrix<Number> &sparse_matrix_host);
+
+    /**
+     * Move constructor. Create a new SparseMatrix by stealing the internal
+     * data.
+     */
+    SparseMatrix(CUDAWrappers::SparseMatrix<Number> &&);
+
+    /**
+     * Copy constructor id deleted.
+     */
+    SparseMatrix(const CUDAWrappers::SparseMatrix<Number> &) = delete;
+
+    /**
+     * Destructor. Free all memory.
+     */
+    ~SparseMatrix();
+
+    /**
+     * Reinitialize the sparse matrix. The sparse matrix on the host is copied
+     * on the device and the elementes are reordered according to the format
+     * supported by cuSPARSE.
+     */
+    void reinit(cusparseHandle_t handle,
+                const ::dealii::SparseMatrix<Number> &sparse_matrix_host);
+    //@}
+
+    /**
+     * @name Information on the matrix
+     */
+    //@{
+    /**
+     * Return the dimension of the codomain (or range) space. Note that the
+     * matrix is of dimension $m \times n$.
+     */
+    size_type m() const;
+
+    /**
+     * Return the dimension of the domain space. Note that the matrix is of
+     * dimension $m \times n$.
+     */
+    size_type n() const;
+
+    /**
+     * Return the number of nonzero elements of this matrix. Actually, it
+     * returns the number of entries in the sparsity pattern; if any of the
+     * entries should happen to be zero, it is counted anyway.
+     */
+    std::size_t n_nonzero_elements() const;
+    //@}
+
+    /**
+     * @name Modifying entries
+     */
+    //@{
+    /**
+     * Multiply the entire matrix by a fixed factor.
+     */
+    SparseMatrix &operator*= (const Number factor);
+
+    /**
+     * Divide the entrie matrix by a fixed factor.
+     */
+    SparseMatrix &operator/= (const Number factor);
+    //@}
+
+    /**
+     * @name Multiplications
+     */
+    //@{
+    /**
+     * Matrix-vector multiplication: let <i>dst = M*src</i> with <i>M</i>
```

Comment:
that's probably a copy-paste thing, but we may as well use LaTeX in doxygen.
(same below)